library(magrittr)
library(tools)
library(stringr)
library(RColorBrewer)
library(readxl)
library(MASS)
library(Matrix)
library(reshape2)
library(gplots)
library(devtools)
library(plyr)
library(Biostrings)
library(muscle)
library(tibble)
library(ShortRead)
library(tibble)
library(tidyverse)
library(vroom)

StartTime <- Sys.time()

###############
## Functions ##
###############

'%nin%' <- Negate('%in%')

uLength <- function(x) {x %>% unique() %>% length()}

#####################
## Get fasta files ##
#####################
DirIn <- "/Users/rtearle/Documents/Roseworthy_Projects/5.Other_Projects/Salmonella/2019_Aug/Salmonella_Concatenated_Fasta_Files"
# These files have the chromosome contigs concatenated, separated by 20Ns

# KC_109
File <- "KC_109.fasta"
KC_109_Seq <- readFasta(file.path(DirIn, File))

Read <- sread(KC_109_Seq) %>% as.character() %>% DNAString()
ReadC <- reverseComplement(Read)

# 5' primer for SSTR6
Primer5 <- "TCGGGCATGCGTTGAAAA"
matchPattern(Primer5, Read)
# views: NONE
matchPattern(Primer5, ReadC)
# views:
#       start     end width
# [1] 1521068 1521085    18 [TCGGGCATGCGTTGAAAA]

# 3' primer for SSTR6
Primer3 <- "GCCAGTCATTCTCCCCACCAG"
matchPattern(Primer3, ReadC)
#views:
#       start     end width
# [1] 1521371 1521391    21 [GCCAGTCATTCTCCCCACCAG]

SSTR6 <- ReadC[1521068:1521391]
SSTR6 %>% as.character()
# "TCGGGCATGCGTTGAAAACCTGCAAAAAAAGAAAAATATCGATGCTTATTATTTTTTCTTTAATTAAATTTTCGCTCAACAAACTTAATTATTAATCCAATGACAGTGAAGTGTGAACTATGCTGAGCATAAACGACATCAATAGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAATCTGAGTATTTACAGGTGAAATTATGAAAAATAGTATATTTTTACCATTGCTTCTGGTGCTGTCGGCAACGACCTTTTCGGCATTGGTGATGGCTGCCAGTCATTCTCCCCACCAG"

Repeat <- "GCAAGG"
matchPattern(Repeat, SSTR6)
# views:
#     start end width
#  [1]   145 150     6 [GCAAGG]
#  [2]   151 156     6 [GCAAGG]
#  [3]   157 162     6 [GCAAGG]
#  [4]   163 168     6 [GCAAGG]
#  [5]   169 174     6 [GCAAGG]
#  [6]   175 180     6 [GCAAGG]
#  [7]   181 186     6 [GCAAGG]
#  [8]   187 192     6 [GCAAGG]
#  [9]   193 198     6 [GCAAGG]
# [10]   199 204     6 [GCAAGG]

# T
File <- "T.fasta"
T_Seq <- readFasta(file.path(DirIn, File))

Read <- sread(T_Seq) %>% as.character() %>% DNAString()
ReadC <- reverseComplement(Read)

Primer5 <- "TCGGGCATGCGTTGAAAA"
matchPattern(Primer5, Read)
# views: NONE
matchPattern(Primer5, ReadC)
# views:
#       start     end width
# [1] 1521131 1521148    18 [TCGGGCATGCGTTGAAAA]

Primer3 <- "GCCAGTCATTCTCCCCACCAG"
matchPattern(Primer3, ReadC)
#views:
#       start     end width
# [1] 1521422 1521442    21 [GCCAGTCATTCTCCCCACCAG]

SSTR6 <- ReadC[1521131:1521442]
SSTR6 %>% as.character()
# "TCGGGCATGCGTTGAAAACCTGCAAAAAAAGAAAAATATCGATGCTTATTATTTTTTCTTTAATTAAATTTTCGCTCAACAAACTTAATTATTAATCCAATGACAGTGAAGTGTGAACTATGCTGAGCATAAACGACATCAATAGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAATCTGAGTATTTACAGGTGAAATTATGAAAAATAGTATATTTTTACCATTGCTTCTGGTGCTGTCGGCAACGACCTTTTCGGCATTGGTGATGGCTGCCAGTCATTCTCCCCACCAG"

Repeat <- "GCAAGG"
matchPattern(Repeat, SSTR6)
# views:
#     start end width
# [1]   145 150     6 [GCAAGG]
# [2]   151 156     6 [GCAAGG]
# [3]   157 162     6 [GCAAGG]
# [4]   163 168     6 [GCAAGG]
# [5]   169 174     6 [GCAAGG]
# [6]   175 180     6 [GCAAGG]
# [7]   181 186     6 [GCAAGG]
# [8]   187 192     6 [GCAAGG]

# TM
File <- "TM.fasta"
TM_Seq <- readFasta(file.path(DirIn, File))

Read <- sread(TM_Seq) %>% as.character() %>% DNAString()
ReadC <- reverseComplement(Read)

Primer5 <- "TCGGGCATGCGTTGAAAA"
matchPattern(Primer5, Read)
# views:
#       start     end width
# [1] 3423701 3423718    18 [TCGGGCATGCGTTGAAAA]
matchPattern(Primer5, ReadC)
# views: NONE

Primer3 <- "GCCAGTCATTCTCCCCACCAG"
matchPattern(Primer3, Read)
#views:
#       start     end width
# [1] 3423992 3424012    21 [GCCAGTCATTCTCCCCACCAG]
matchPattern(Primer3, ReadC)
# views: NONE

SSTR6 <- Read[3423701:3424012]
SSTR6 %>% as.character()
# "TCGGGCATGCGTTGAAAACCTGCAAAAAAAGAAAAATATCGATGCTTATTATTTTTTCTTTAATTAAATTTTCGCTCAACAAACTTAATTATTAATCCAATGACAGTGAAGTGTGAACTATGCTGAGCATAAACGACATCAATAGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAATCTGAGTATTTACAGGTGAAATTATGAAAAATAGTATATTTTTACCATTGCTTCTGGTGCTGTCGGCAACGACCTTTTCGGCATTGGTGATGGCTGCCAGTCATTCTCCCCACCAG"

Repeat <- "GCAAGG"
matchPattern(Repeat, SSTR6)
# views:
#     start end width
# [1]   145 150     6 [GCAAGG]
# [2]   151 156     6 [GCAAGG]
# [3]   157 162     6 [GCAAGG]
# [4]   163 168     6 [GCAAGG]
# [5]   169 174     6 [GCAAGG]
# [6]   175 180     6 [GCAAGG]
# [7]   181 186     6 [GCAAGG]
# [8]   187 192     6 [GCAAGG]

# TMu
File <- "TMu.fasta"
TMu_Seq <- readFasta(file.path(DirIn, File))

Read <- sread(TMu_Seq) %>% as.character() %>% DNAString()
ReadC <- reverseComplement(Read)

Primer5 <- "TCGGGCATGCGTTGAAAA"
matchPattern(Primer5, Read)
# views: NONE
matchPattern(Primer5, ReadC)
# views:
#       start     end width
# [1521133 1521150    18 [TCGGGCATGCGTTGAAAA]

Primer3 <- "GCCAGTCATTCTCCCCACCAG"
matchPattern(Primer3, ReadC)
#views:
#       start     end width
# [1] 1521436 1521456    21 [GCCAGTCATTCTCCCCACCAG]
matchPattern(Primer3, Read)
# views: NONE

SSTR6 <- ReadC[1521133:1521456]
SSTR6 %>% as.character()
# "TCGGGCATGCGTTGAAAACCTGCAAAAAAAGAAAAATATCGATGCTTATTATTTTTTCTTTAATTAAATTTTCGCTCAACAAACTTAATTATTAATCCAATGACAGTGAAGTGTGAACTATGCTGAGCATAAACGACATCAATAGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAATCTGAGTATTTACAGGTGAAATTATGAAAAATAGTATATTTTTACCATTGCTTCTGGTGCTGTCGGCAACGACCTTTTCGGCATTGGTGATGGCTGCCAGTCATTCTCCCCACCAG"

Repeat <- "GCAAGG"
matchPattern(Repeat, SSTR6)
# views:
#      start end width
#  [1]   145 150     6 [GCAAGG]
#  [2]   151 156     6 [GCAAGG]
#  [3]   157 162     6 [GCAAGG]
#  [4]   163 168     6 [GCAAGG]
#  [5]   169 174     6 [GCAAGG]
#  [6]   175 180     6 [GCAAGG]
#  [7]   181 186     6 [GCAAGG]
#  [8]   187 192     6 [GCAAGG]
#  [9]   193 198     6 [GCAAGG]
# [10]   199 204     6 [GCAAGG]

# "TCGGGCATGCGTTGAAAACCTGCAAAAAAAGAAAAATATCGATGCTTATTATTTTTTCTTTAATTAAATTTTCGCTCAACAAACTTAATTATTAATCCAATGACAGTGAAGTGTGAACTATGCTGAGCATAAACGACATCAATAGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAATCTGAGTATTTACAGGTGAAATTATGAAAAATAGTATATTTTTACCATTGCTTCTGGTGCTGTCGGCAACGACCTTTTCGGCATTGGTGATGGCTGCCAGTCATTCTCCCCACCAG"
# "TCGGGCATGCGTTGAAAACCTGCAAAAAAAGAAAAATATCGATGCTTATTATTTTTTCTTTAATTAAATTTTCGCTCAACAAACTTAATTATTAATCCAATGACAGTGAAGTGTGAACTATGCTGAGCATAAACGACATCAATAGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAATCTGAGTATTTACAGGTGAAATTATGAAAAATAGTATATTTTTACCATTGCTTCTGGTGCTGTCGGCAACGACCTTTTCGGCATTGGTGATGGCTGCCAGTCATTCTCCCCACCAG"
# "TCGGGCATGCGTTGAAAACCTGCAAAAAAAGAAAAATATCGATGCTTATTATTTTTTCTTTAATTAAATTTTCGCTCAACAAACTTAATTATTAATCCAATGACAGTGAAGTGTGAACTATGCTGAGCATAAACGACATCAATAGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAATCTGAGTATTTACAGGTGAAATTATGAAAAATAGTATATTTTTACCATTGCTTCTGGTGCTGTCGGCAACGACCTTTTCGGCATTGGTGATGGCTGCCAGTCATTCTCCCCACCAG"
# "TCGGGCATGCGTTGAAAACCTGCAAAAAAAGAAAAATATCGATGCTTATTATTTTTTCTTTAATTAAATTTTCGCTCAACAAACTTAATTATTAATCCAATGACAGTGAAGTGTGAACTATGCTGAGCATAAACGACATCAATAGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAATCTGAGTATTTACAGGTGAAATTATGAAAAATAGTATATTTTTACCATTGCTTCTGGTGCTGTCGGCAACGACCTTTTCGGCATTGGTGATGGCTGCCAGTCATTCTCCCCACCAG"

doMuscleComp <- function (Strings, Names, Quiet, Type) {
    if (Type == "AA") {
        Strings <- Biostrings::AAStringSet(Strings)
    } else if (Type == "DNA") {
        Strings <- Biostrings::DNAStringSet(Strings)
    } else {
        stop(sprintf("Expected type of 'AA' or 'DNA', got '%s'", Type))
    }
    names(Strings) <- Names
    Align <- muscle::muscle(Strings, quiet = "Quiet")
}

Seqs <- c("TCGGGCATGCGTTGAAAACCTGCAAAAAAAGAAAAATATCGATGCTTATTATTTTTTCTTTAATTAAATTTTCGCTCAACAAACTTAATTATTAATCCAATGACAGTGAAGTGTGAACTATGCTGAGCATAAACGACATCAATAGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAATCTGAGTATTTACAGGTGAAATTATGAAAAATAGTATATTTTTACCATTGCTTCTGGTGCTGTCGGCAACGACCTTTTCGGCATTGGTGATGGCTGCCAGTCATTCTCCCCACCAG",
          "TCGGGCATGCGTTGAAAACCTGCAAAAAAAGAAAAATATCGATGCTTATTATTTTTTCTTTAATTAAATTTTCGCTCAACAAACTTAATTATTAATCCAATGACAGTGAAGTGTGAACTATGCTGAGCATAAACGACATCAATAGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAATCTGAGTATTTACAGGTGAAATTATGAAAAATAGTATATTTTTACCATTGCTTCTGGTGCTGTCGGCAACGACCTTTTCGGCATTGGTGATGGCTGCCAGTCATTCTCCCCACCAG",
          "TCGGGCATGCGTTGAAAACCTGCAAAAAAAGAAAAATATCGATGCTTATTATTTTTTCTTTAATTAAATTTTCGCTCAACAAACTTAATTATTAATCCAATGACAGTGAAGTGTGAACTATGCTGAGCATAAACGACATCAATAGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAATCTGAGTATTTACAGGTGAAATTATGAAAAATAGTATATTTTTACCATTGCTTCTGGTGCTGTCGGCAACGACCTTTTCGGCATTGGTGATGGCTGCCAGTCATTCTCCCCACCAG",
          "TCGGGCATGCGTTGAAAACCTGCAAAAAAAGAAAAATATCGATGCTTATTATTTTTTCTTTAATTAAATTTTCGCTCAACAAACTTAATTATTAATCCAATGACAGTGAAGTGTGAACTATGCTGAGCATAAACGACATCAATAGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAAGGGCAATCTGAGTATTTACAGGTGAAATTATGAAAAATAGTATATTTTTACCATTGCTTCTGGTGCTGTCGGCAACGACCTTTTCGGCATTGGTGATGGCTGCCAGTCATTCTCCCCACCAG"
          )

Seq2 <- DNAStringSet(Seqs)

Align <- muscle::muscle(Seq2)
AlignM <- as.matrix(Align)

Align %>% as.character()

